src/main/java/org/openstreetmap/atlas/geography/atlas/BareAtlas.java

```diff
@@ -914,6 +942,34 @@ private void checkRelationMembersAndIndexRelation(final Relation relation,
         }
     }
 
+    private <Member extends AtlasEntity> Supplier<Iterable<Member>> getCachingSupplier(
+            final Iterable<Member> source, final ItemType type)
+    {
+        final List<Set<Long>> memberIdentifiersIntersecting = new ArrayList<>();
+        // A supplier that will effectively cache all the members intersecting that polygon. This is
+        // thread safe because the cache is internal to the supplier's scope.
+        @SuppressWarnings("unchecked")
+        final Supplier<Iterable<Member>> result = () ->
+        {
+            if (memberIdentifiersIntersecting.isEmpty())
+            {
+                final Set<Long> intersecting = new HashSet<>();
+                memberIdentifiersIntersecting.add(intersecting);
+                return Iterables.stream(source).map(member ->
```

Comment:
I find it a little confusing that the call to `map` does not actually perform a mapping operation, and instead produces a side effect on some external global state. I think you are doing this to avoid calling `Iterables.stream()` twice (Once for a `forEach` and once for the `collect`). So here, you are using `map` as a modified `forEach` that returns an identical stream (so that `collect` can be called in the same line)?

If this is the case, maybe leave a comment that explains this, since this use of `map` is bit non-idiomatic :)

Reply:
+1 to this

Reply:
That works :) I will add a comment.